pennylane/qchem/vibrational/bosonic.py

```diff
@@ -0,0 +1,624 @@
+            raise ValueError(
+                f"Arithmetic Bose operations can only accept an array of length 1, "
+                f"but received {other} of length {len(other)}"
+            )
+
+        return BoseSentence({self: other})
+
+    def __rmul__(self, other):
+        r"""Reverse multiply a BoseWord
+
+        Multiplies a BoseWord "from the left" with an object that can't be modified
+        to support __mul__ for BoseWord. Will be defaulted in for example
+        ``2 * BoseWord({(0, 0): "+"})``, where the ``__mul__`` operator on an integer
+        will fail to multiply with a BoseWord"""
+
+        return self.__mul__(other)
+
+    def __pow__(self, value):
+        r"""Exponentiate a Bose word to an integer power.
+
+        >>> w = BoseWord({(0, 0) : '+', (1, 1) : '-'})
+        >>> print(w**3)
+        b⁺(0) b(1) b⁺(0) b(1) b⁺(0) b(1)
+        """
+
+        if value < 0 or not isinstance(value, int):
+            raise ValueError("The exponent must be a positive integer.")
+
+        operator = BoseWord({})
+
+        for _ in range(value):
+            operator *= self
+
+        return operator
+
+    def normal_order(self):
+        r"""Convert a BoseWord to its normal-ordered form.
+
+        >>> bw = BoseWord({(0, 0): "-", (1, 0): "-", (2, 0): "+", (3, 0): "+"})
+        >>> print(bw.normal_order())
+        4.0 * b⁺(0) b(0)
+        + 2.0 * I
+        + 1.0 * b⁺(0) b⁺(0) b(0) b(0)
+        """
+        bw_terms = sorted(self)
+        len_op = len(bw_terms)
+        bw_comm = BoseSentence({BoseWord({}): 0.0}, is_hardcore=self.is_hardcore)
+
+        if len_op == 0:
+            return 1 * BoseWord({})
+
+        bw = self
+
+        l = 0
+        for r in range(len_op):
+            if self[bw_terms[r]] == "+":
+                if l == r:
+                    l += 1
+                    continue
+
+                bs = bw.shift_operator(r, l)
+                bs_as_list = sorted(list(bs.items()), key=lambda x: len(x[0].keys()), reverse=True)
+                bw = bs_as_list[0][0]
+                if l > 0:
+                    bw_as_list = sorted(list(bw.keys()))
+                    if bw_as_list[l - 1][1] > bw_as_list[l][1]:
+                        temp_bs = bw.shift_operator(l - 1, l)
+                        bw = list(temp_bs.items())[0][0]
+
+                for i in range(1, len(bs_as_list)):
+                    bw_comm += bs_as_list[i][0] * bs_as_list[i][1]
+
+                l += 1
+
+        ordered_op = bw + bw_comm.normal_order()
+        ordered_op.simplify(tol=1e-8)
+
+        if self.is_hardcore:
+            for bw, _ in ordered_op.items():
+                bw_arr = list(bw.keys())
+                indice_arr = [x[1] for x in bw_arr]
+                if len(indice_arr) != len(set(indice_arr)):
+                    ordered_op[bw] = 0
+
+        ordered_op.simplify(tol=1e-8)
+        return ordered_op
+
+    def shift_operator(self, initial_position, final_position):
```

Comment:
What is the best way to shift an operator multiple time? For example, what is the UI for converting `b(0) b(1) b(2)` to `b(1) b(2) b(0)`?

Reply:
It's the same as in FermiWord. To shift from `b(0) b(1) b(2)`  to `b(1) b(2) b(0)`, you can just do bw.shift_operator(0, 2). You can think of b(0) swapping with b1 first, and then with b2.